In [2]:
import pandas as pd

# Cargar datos
df = pd.read_csv("/content/cine-chile-limpia-oficial.csv", sep=";")


# Echar un vistazo
df.head(5)

nombre-pelicula   año  \
0                          La industria del salitre   1915   
1                          La industria del salitre   1915   
2  La baraja de la muerte (o El enigma de la call...  1916   
3                                El hombre de acero   1917   
4  La agonía de Arauco (o El olvido de los muertos)   1917   

          tipo-largometraje duracion-min tipo-filme                  color  \
0  Largometraje documental            NI     35 mm    Blanco & negro, Muda   
1  Largometraje documental            NI     35 mm    Blanco & negro, Muda   
2  Largometraje de ficción           100     35 mm    Blanco & negro, Muda   
3  Largometraje de ficción            NI     35 mm    Blanco & negro, Muda   
4  Largometraje de ficción            NI     35 mm    Blanco & negro, Muda   

             director-01   idioma consol-idioma lugar-prod  ... Columna231  \
0      Francisco Caamaño  Español       Español      Chile  ...        NaN   
1      Francisco Caamaño  Español       Español      Chile  ...        NaN   
2  Salvador Giambastiani  Español       Español      Chile  ...        NaN   
3    Jorge Délano «Coke»  Español       Español      Chile  ...        NaN   
4     Gabriela Bussenius  Español       Español      Chile  ...        NaN   

  Columna232 Columna233 Columna234 Columna235 Columna236  Columna237  \
0        NaN        NaN        NaN        NaN        NaN         NaN   
1        NaN        NaN        NaN        NaN        NaN         NaN   
2        NaN        NaN        NaN        NaN        NaN         NaN   
3        NaN        NaN        NaN        NaN        NaN         NaN   
4        NaN        NaN        NaN        NaN        NaN         NaN   

   Columna238  Columna239  Columna240  
0         NaN         NaN         NaN  
1         NaN         NaN         NaN  
2         NaN         NaN         NaN  
3         NaN         NaN         NaN  
4         NaN         NaN         NaN  

[5 rows x 256 columns]

In [ ]:
combined_chart.save('evolucion_directores_cine_chileno.html')
print("El gráfico se ha guardado como 'evolucion_directores_cine_chileno.html'")

El gráfico se ha guardado como 'evolucion_directores_cine_chileno.html'


# Task
Analyze the gender representation of film directors over the years from the `/content/cine-chile-limpia-oficial.csv` dataset. Create a new column to identify the gender of each director, then calculate the yearly count of male and female directors, as well as the female-to-male ratio. Finally, visualize these trends using an Altair stacked area chart showing the number of directors by gender, overlaid with a line chart displaying the female-to-male ratio over time.

## Preparar Datos de Género de Directores

### Subtask:
Se analizará la columna 'director-01' para identificar el género (masculino/femenino) de cada director y se creará una nueva columna 'genero_director'. Luego, se agruparán los datos por año y género para contar el número de directores y directoras por cada año.


**Reasoning**:
The first step is to inspect the 'director-01' column to understand its content and format, which is crucial for inferring gender later. Displaying unique values will help identify common naming patterns and potential challenges.



In [29]:
print(df['director-01'].value_counts())

director-01
Antonio de Santos      16
José Bohr              16
Cristián Sánchez G.    16
Raúl Ruiz              15
Alberto Santana        15
                       ..
Diego Céspedes          1
José Mañodes            1
Meliza Luna Venegas     1
Vanessa Miller          1
Manuel Seva             1
Name: count, Length: 1226, dtype: int64


**Reasoning**:
As per the subtask instructions, a new column named 'genero_director' needs to be created in the DataFrame, and it should be initialized with a default value. 'Desconocido' is a suitable default value as it indicates that the gender is yet to be determined.



In [30]:
df['genero_director'] = 'Desconocido'
print("Columna 'genero_director' creada e inicializada.")


Columna 'genero_director' creada e inicializada.


**Reasoning**:
To classify the gender of directors, I need to define lists of common male and female Spanish names. I will create a function that attempts to determine the gender based on the first name of each director. This function will then be applied to the 'director-01' column to populate the 'genero_director' column, marking names not found in either list as 'Desconocido'.



In [31]:
nombres_masculinos = [
    'José', 'Juan', 'Luis', 'Carlos', 'Francisco', 'Manuel', 'Antonio', 'Miguel', 'Pedro', 'Pablo',
    'Fernando', 'Alejandro', 'Andrés', 'David', 'Jorge', 'Daniel', 'Javier', 'Sergio', 'Ricardo', 'Alberto',
    'Eduardo', 'Enrique', 'Rafael', 'Diego', 'Gonzalo', 'Héctor', 'Mario', 'Martín', 'Oscar', 'Ramón',
    'Roberto', 'Rubén', 'Santiago', 'Sebastián', 'Vicente', 'Esteban', 'Álvaro', 'Gabriel', 'Guillermo', 'Ignacio',
    'Joaquín', 'Leonardo', 'Marcos', 'Nicolás', 'Raúl', 'Rodrigo', 'Adrián', 'Agustín', 'Alfredo', 'Benjamín',
    'Cristián', 'Felipe', 'Germán', 'Gustavo', 'Patricio', 'Víctor'
]
nombres_femeninos = [
    'María', 'Ana', 'Laura', 'Sofía', 'Valentina', 'Camila', 'Isabel', 'Paula', 'Daniela', 'Carolina',
    'Andrea', 'Elena', 'Gabriela', 'Jimena', 'Lucía', 'Mariana', 'Natalia', 'Pilar', 'Sara', 'Silvia',
    'Verónica', 'Alejandra', 'Beatriz', 'Carmen', 'Clara', 'Diana', 'Dolores', 'Emilia', 'Eva', 'Florencia',
    'Gloria', 'Irene', 'Julia', 'Lorena', 'Magdalena', 'Marta', 'Mónica', 'Patricia', 'Raquel', 'Rocío',
    'Rosa', 'Sandra', 'Teresa', 'Victoria', 'Ximena', 'Cecilia', 'Consuelo', 'Francisca', 'Jessica', 'Katherine',
    'Claudia', 'Elizabeth', 'Macarena', 'Melissa', 'Romina', 'Soledad'
]

def inferir_genero(nombre_completo):
    if pd.isna(nombre_completo):
        return 'Desconocido'
    # Convertir a minúsculas para una comparación sin distinción de mayúsculas y minúsculas
    nombre_completo_lower = str(nombre_completo).lower()

    # Separar el nombre en palabras y tomar la primera como posible nombre propio
    primer_nombre = nombre_completo_lower.split(' ')[0]

    # Buscar nombres femeninos
    for nombre_fem in nombres_femeninos:
        if primer_nombre == nombre_fem.lower():
            return 'Femenino'

    # Buscar nombres masculinos
    for nombre_masc in nombres_masculinos:
        if primer_nombre == nombre_masc.lower():
            return 'Masculino'

    return 'Desconocido'

df['genero_director'] = df['director-01'].apply(inferir_genero)

print("La columna 'genero_director' ha sido actualizada con el género inferido.")
print(df['genero_director'].value_counts())

La columna 'genero_director' ha sido actualizada con el género inferido.
genero_director
Masculino      1106
Desconocido     868
Femenino        145
Name: count, dtype: int64


**Reasoning**:
Now that the 'genero_director' column has been populated, the next step is to group the DataFrame by 'año' and 'genero_director' and then count the number of unique directors for each group, as specified in the subtask. This will provide the annual count of male and female directors.



In [32]:
directores_por_ano_genero = df.groupby(['año', 'genero_director'])['director-01'].nunique().reset_index()

print("Conteo de directores únicos por año y género:")
print(directores_por_ano_genero.head())


Conteo de directores únicos por año y género:
    año genero_director  director-01
0  1915       Masculino            1
1  1916     Desconocido            1
2  1917     Desconocido            1
3  1917        Femenino            1
4  1917       Masculino            1


## Calcular la Tasa de Mujeres por Hombre

### Subtask:
Utilizando los conteos de directores y directoras por año, se calculará la tasa de mujeres por hombre (F/M) para cada año.


**Reasoning**:
To calculate the female-to-male ratio, I need to reshape the `directores_por_ano_genero` DataFrame so that each year has separate columns for 'Femenino', 'Masculino', and 'Desconocido' director counts. Pivoting will achieve this, and filling `NaN` values with 0 ensures that years without a particular gender count are handled correctly for subsequent calculations.



In [33]:
if 'directores_por_ano_genero' in locals():
    directores_pivot = directores_por_ano_genero.pivot_table(
        index='año',
        columns='genero_director',
        values='director-01',
        fill_value=0
    ).reset_index()

    # Ensure all gender columns exist, filling with 0 if not present in the pivot
    # Now, we expect only 'Femenino' and 'Masculino' columns.
    for col in ['Femenino', 'Masculino']:
        if col not in directores_pivot.columns:
            directores_pivot[col] = 0

    print("DataFrame pivotado por año y género (solo Femenino/Masculino):")
    print(directores_pivot.head())
else:
    print("No se ha definido 'directores_por_ano_genero', no se puede pivotar el DataFrame.")

DataFrame pivotado por año y género (solo Femenino/Masculino):
genero_director   año  Desconocido  Femenino  Masculino
0                1915          0.0       0.0        1.0
1                1916          1.0       0.0        0.0
2                1917          1.0       1.0        1.0
3                1918          1.0       0.0        0.0
4                1919          0.0       0.0        1.0


**Reasoning**:
Now that the data is pivoted, I need to calculate the female-to-male ratio. This requires dividing the 'Femenino' column by the 'Masculino' column, and it's crucial to handle cases where 'Masculino' is zero to avoid division errors and replace resulting infinite or NaN values with 0 to ensure the ratio is well-defined.



In [34]:
if 'directores_pivot' in locals():
    directores_pivot['ratio_F_M'] = directores_pivot['Femenino'] / directores_pivot['Masculino']
    directores_pivot['ratio_F_M'] = directores_pivot['ratio_F_M'].replace([float('inf'), -float('inf')], 0).fillna(0)

    print("DataFrame con la tasa de mujeres por hombres (solo Femenino/Masculino):")
    print(directores_pivot.head())
else:
    print("No se ha definido 'directores_pivot', no se puede calcular la tasa F/M.")

DataFrame con la tasa de mujeres por hombres (solo Femenino/Masculino):
genero_director   año  Desconocido  Femenino  Masculino  ratio_F_M
0                1915          0.0       0.0        1.0        0.0
1                1916          1.0       0.0        0.0        0.0
2                1917          1.0       1.0        1.0        1.0
3                1918          1.0       0.0        0.0        0.0
4                1919          0.0       0.0        1.0        0.0


## Crear Gráfico de Área Apilado y Línea

### Subtask:
Se generará un gráfico de Altair que combine un gráfico de área apilado para mostrar la cantidad de directores y directoras a lo largo de los años, superponiendo un gráfico de línea con un eje secundario para visualizar la tasa de mujeres por hombre.


**Reasoning**:
To prepare the data for the stacked area chart, I need to 'melt' the `directores_pivot` DataFrame. This will transform the gender columns ('Femenino', 'Masculino', 'Desconocido') into rows, making the data suitable for Altair's stacking capabilities.



In [7]:
import altair as alt

df_melted = directores_pivot.melt(
    id_vars=['año'],
    value_vars=['Femenino', 'Masculino'], # Se excluye 'Desconocido' aquí
    var_name='genero_director',
    value_name='count'
)

print("DataFrame 'df_melted' creado y listo para la visualización (sin 'Desconocido'):")
print(df_melted.head())

NameError: name 'directores_pivot' is not defined

**Reasoning**:
Now that the data is prepared in `df_melted` for the stacked area chart and `directores_pivot` for the ratio line chart, I will create both charts, combine them using `alt.layer`, assign appropriate axes, tooltips, and titles, and then display the final visualization as per the subtask instructions.



In [13]:
base = alt.Chart(df_melted).encode(
    x=alt.X('año:O', axis=alt.Axis(title='Año')),
    tooltip=['año', 'genero_director', 'count']
)

# Gráfico de área apilado para el conteo de directores por género
area_chart = base.mark_area().encode(
    y=alt.Y('count:Q', axis=alt.Axis(title='Número de Directores')),
    color=alt.Color('genero_director:N', title='Género del Director',
                    scale=alt.Scale(domain=['Femenino', 'Masculino'], range=['#88937B', '#5A0F05'])) # Colores actualizados
)

# Gráfico de línea para la tasa F/M
line_chart = alt.Chart(directores_pivot).mark_line(color='red').encode(
    x=alt.X('año:O'),
    y=alt.Y('ratio_F_M:Q', axis=alt.Axis(title='Ratio Femenino/Masculino', titleColor='red')),
    tooltip=['año', alt.Tooltip('ratio_F_M', format='.2f', title='Ratio F/M')]
)

# Combinar los gráficos
combined_chart = alt.layer(area_chart, line_chart).resolve_scale(
    y='independent'  # Permite que ambos gráficos tengan sus propios ejes Y
).properties(
    title='Evolución de la Participación de Directores por Género y Ratio F/M en el Cine Chileno'
).interactive()

combined_chart

alt.LayerChart(...)

In [14]:
combined_chart.save('evolucion_directores_cine_chileno.html')
print("El gráfico con los colores actualizados se ha guardado como 'evolucion_directores_cine_chileno.html'")

El gráfico con los colores actualizados se ha guardado como 'evolucion_directores_cine_chileno.html'


In [12]:
import pandas as pd
import os
import altair as alt

file_path = "/content/cine-chile-limpia-oficial.csv"

# Check if the file exists and load data
if not os.path.exists(file_path):
    print(f"Error: El archivo '{file_path}' no se encontró. Por favor, asegúrate de subir el archivo CSV a esta ubicación.")
else:
    # Load df if it's not already loaded or is empty
    if 'df' not in locals() or df.empty:
        df = pd.read_csv(file_path, sep=";")

    # Define gender inference components if not already defined
    if 'nombres_masculinos' not in locals() or 'nombres_femeninos' not in locals() or 'inferir_genero' not in locals():
        nombres_masculinos = [
            'José', 'Juan', 'Luis', 'Carlos', 'Francisco', 'Manuel', 'Antonio', 'Miguel', 'Pedro', 'Pablo',
            'Fernando', 'Alejandro', 'Andrés', 'David', 'Jorge', 'Daniel', 'Javier', 'Sergio', 'Ricardo', 'Alberto',
            'Eduardo', 'Enrique', 'Rafael', 'Diego', 'Gonzalo', 'Héctor', 'Mario', 'Martín', 'Oscar', 'Ramón',
            'Roberto', 'Rubén', 'Santiago', 'Sebastián', 'Vicente', 'Esteban', 'Álvaro', 'Gabriel', 'Guillermo', 'Ignacio',
            'Joaquín', 'Leonardo', 'Marcos', 'Nicolás', 'Raúl', 'Rodrigo', 'Adrián', 'Agustín', 'Alfredo', 'Benjamín',
            'Cristián', 'Felipe', 'Germán', 'Gustavo', 'Patricio', 'Víctor'
        ]
        nombres_femeninos = [
            'María', 'Ana', 'Laura', 'Sofía', 'Valentina', 'Camila', 'Isabel', 'Paula', 'Daniela', 'Carolina',
            'Andrea', 'Elena', 'Gabriela', 'Jimena', 'Lucía', 'Mariana', 'Natalia', 'Pilar', 'Sara', 'Silvia',
            'Verónica', 'Alejandra', 'Beatriz', 'Carmen', 'Clara', 'Diana', 'Dolores', 'Emilia', 'Eva', 'Florencia',
            'Gloria', 'Irene', 'Julia', 'Lorena', 'Magdalena', 'Marta', 'Mónica', 'Patricia', 'Raquel', 'Rocío',
            'Rosa', 'Sandra', 'Teresa', 'Victoria', 'Ximena', 'Cecilia', 'Consuelo', 'Francisca', 'Jessica', 'Katherine',
            'Claudia', 'Elizabeth', 'Macarena', 'Melissa', 'Romina', 'Soledad'
        ]

        def inferir_genero(nombre_completo):
            if pd.isna(nombre_completo):
                return 'Desconocido'
            nombre_completo_lower = str(nombre_completo).lower()
            primer_nombre = nombre_completo_lower.split(' ')[0]
            for nombre_fem in nombres_femeninos:
                if primer_nombre == nombre_fem.lower():
                    return 'Femenino'
            for nombre_masc in nombres_masculinos:
                if primer_nombre == nombre_masc.lower():
                    return 'Masculino'
            return 'Desconocido'
        df['genero_director'] = df['director-01'].apply(inferir_genero)

    # Filter out 'Desconocido' entries if df is available
    if 'df' in locals() and not df.empty:
        df_filtrado = df[df['genero_director'] != 'Desconocido'].copy() # Use .copy() to avoid SettingWithCopyWarning
        print("DataFrame filtrado para excluir directores 'Desconocido'.")
        print(df_filtrado['genero_director'].value_counts())

        # Re-calculate directores_por_ano_genero using the filtered DataFrame
        directores_por_ano_genero = df_filtrado.groupby(['año', 'genero_director'])['director-01'].nunique().reset_index()

        print("Conteo de directores únicos por año y género (solo Femenino/Masculino):")
        print(directores_por_ano_genero.head())

        # Pivot the data for visualization
        directores_pivot = directores_por_ano_genero.pivot_table(
            index='año',
            columns='genero_director',
            values='director-01',
            fill_value=0
        ).reset_index()

        # Ensure all gender columns exist, filling with 0 if not present
        for col in ['Femenino', 'Masculino']:
            if col not in directores_pivot.columns:
                directores_pivot[col] = 0

        # Calculate ratio F/M
        directores_pivot['ratio_F_M'] = directores_pivot['Femenino'] / directores_pivot['Masculino']
        directores_pivot['ratio_F_M'] = directores_pivot['ratio_F_M'].replace([float('inf'), -float('inf')], 0).fillna(0)

        # Melt the data for stacked area chart
        df_melted = directores_pivot.melt(
            id_vars=['año'],
            value_vars=['Femenino', 'Masculino'],
            var_name='genero_director',
            value_name='count'
        )

        print("DataFrame 'df_melted' creado y listo para la visualización (sin 'Desconocido'):")
        print(df_melted.head())

        # Create the combined chart
        base = alt.Chart(df_melted).encode(
            x=alt.X('año:O', axis=alt.Axis(title='Año')),
            tooltip=['año', 'genero_director', 'count']
        )

        area_chart = base.mark_area().encode(
            y=alt.Y('count:Q', axis=alt.Axis(title='Número de Directores')),
            color=alt.Color('genero_director:N', title='Género del Director',
                            scale=alt.Scale(domain=['Femenino', 'Masculino'], range=['#FFC0CB', '#ADD8E6']))
        )

        line_chart = alt.Chart(directores_pivot).mark_line(color='red').encode(
            x=alt.X('año:O'),
            y=alt.Y('ratio_F_M:Q', axis=alt.Axis(title='Ratio Femenino/Masculino', titleColor='red')),
            tooltip=['año', alt.Tooltip('ratio_F_M', format='.2f', title='Ratio F/M')]
        )

        combined_chart = alt.layer(area_chart, line_chart).resolve_scale(
            y='independent'
        ).properties(
            title='Evolución de la Participación de Directores por Género y Ratio F/M en el Cine Chileno'
        ).interactive()

        # Display and save the chart
        combined_chart.save('evolucion_directores_cine_chileno.html')
        print("El gráfico de la evolución de la participación de directores por género se ha guardado como 'evolucion_directores_cine_chileno.html'")

    else:
        print("No se pudo cargar el DataFrame 'df', no se realizó el filtrado ni el conteo.")

DataFrame filtrado para excluir directores 'Desconocido'.
genero_director
Masculino    1106
Femenino      145
Name: count, dtype: int64
Conteo de directores únicos por año y género (solo Femenino/Masculino):
    año genero_director  director-01
0  1915       Masculino            1
1  1917        Femenino            1
2  1917       Masculino            1
3  1919       Masculino            1
4  1920       Masculino            2
DataFrame 'df_melted' creado y listo para la visualización (sin 'Desconocido'):
    año genero_director  count
0  1915        Femenino    0.0
1  1917        Femenino    1.0
2  1919        Femenino    0.0
3  1920        Femenino    0.0
4  1921        Femenino    0.0
El gráfico de la evolución de la participación de directores por género se ha guardado como 'evolucion_directores_cine_chileno.html'


In [10]:
import altair as alt

base = alt.Chart(df_melted).encode(
    x=alt.X('año:O', axis=alt.Axis(title='Año')),
    tooltip=['año', 'genero_director', 'count']
)

# Gráfico de área apilado para el conteo de directores por género
area_chart = base.mark_area().encode(
    y=alt.Y('count:Q', axis=alt.Axis(title='Número de Directores')),
    color=alt.Color('genero_director:N', title='Género del Director',
                    scale=alt.Scale(domain=['Femenino', 'Masculino'], range=['#FFC0CB', '#ADD8E6']))
)

# Gráfico de línea para la tasa F/M
line_chart = alt.Chart(directores_pivot).mark_line(color='red').encode(
    x=alt.X('año:O'),
    y=alt.Y('ratio_F_M:Q', axis=alt.Axis(title='Ratio Femenino/Masculino', titleColor='red')),
    tooltip=['año', alt.Tooltip('ratio_F_M', format='.2f', title='Ratio F/M')]
)

# Combinar los gráficos
combined_chart = alt.layer(area_chart, line_chart).resolve_scale(
    y='independent'  # Permite que ambos gráficos tengan sus propios ejes Y
).properties(
    title='Evolución de la Participación de Directores por Género y Ratio F/M en el Cine Chileno'
).interactive()

combined_chart

NameError: name 'df_melted' is not defined

In [11]:
combined_chart.save('evolucion_directores_cine_chileno.html')
print("El gráfico de la evolución de la participación de directores por género se ha guardado como 'evolucion_directores_cine_chileno.html'")

NameError: name 'combined_chart' is not defined

In [3]:
combined_chart.save('evolucion_directores_cine_chileno.html')
print("El gráfico de la evolución de la participación de directores por género se ha guardado como 'evolucion_directores_cine_chileno.html'")

NameError: name 'combined_chart' is not defined

In [4]:
combined_chart.save('evolucion_directores_cine_chileno.html')
print("El gráfico se ha guardado como 'evolucion_directores_cine_chileno.html'")

NameError: name 'combined_chart' is not defined

## Visualización de los 20 Directores con Más Películas

### Subtask:
Se identificará a los 20 directores con la mayor cantidad de películas y se creará un gráfico de barras horizontal que muestre este listado, utilizando una paleta de colores personalizada.

**Reasoning**:
To identify the top 20 directors, I need to count the number of unique films associated with each director from the `df` DataFrame. Grouping by 'director-01' and counting 'nombre-pelicula' will provide this, and then sorting and selecting the top 20 will give the required data for the visualization.

In [49]:
directores_conteo = df.groupby('director-01')['nombre-pelicula'].nunique().reset_index()
directores_conteo = directores_conteo.rename(columns={'nombre-pelicula': 'cantidad_peliculas'})

top_20_directores = directores_conteo.sort_values(by='cantidad_peliculas', ascending=False).head(20)

print("Los 20 directores con la mayor cantidad de películas:")
print(top_20_directores)

Los 20 directores con la mayor cantidad de películas:
                 director-01  cantidad_peliculas
283      Cristián Sánchez G.                  16
117        Antonio de Santos                  16
663                José Bohr                  16
17           Alberto Santana                  15
1038               Raúl Ruiz                  15
954            Pablo Larraín                  12
673   José Luis Torres Leiva                  12
994      Patricio Valladares                  12
644      Jorge Délano «Coke»                  11
912            Nicolás López                  11
530       Gonzalo Justiniano                  10
986          Patricio Guzmán                  10
577           Ignacio Agüero                  10
558          Hermann Mondaca                  10
424       Eugenio de Liguoro                  10
779             Luis R. Vera                   9
886            Miguel Littin                   9
410    Ernesto Díaz Espinoza                   9
143            

**Reasoning**:
Now that I have the data for the top 20 directors, I will create an Altair horizontal bar chart. The chart will display the director names on the Y-axis and the count of films on the X-axis. The bars will be colored using the custom palette provided by the user, and the chart will be interactive for better exploration.

In [50]:
import altair as alt

palette = ["#41402A", "#5A0F05", "#9B4337", "#88937B", "#CFC38D", "#778F3B", "#598299"]

chart_top_directors = alt.Chart(top_20_directores).mark_bar().encode(
    x=alt.X('cantidad_peliculas:Q', title='Cantidad de Películas'),
    y=alt.Y('director-01:N', sort='-x', title='Director'),
    color=alt.Color('director-01:N', scale=alt.Scale(range=palette), legend=None), # Color by director using the palette
    tooltip=['director-01', 'cantidad_peliculas']
).properties(
    title='Top 20 Directores con Mayor Cantidad de Películas'
).interactive()

chart_top_directors

alt.Chart(...)

In [51]:
chart_top_directors.save('top_20_directores.html')
print("El gráfico de los Top 20 Directores se ha guardado como 'top_20_directores.html'")

El gráfico de los Top 20 Directores se ha guardado como 'top_20_directores.html'


**Reasoning**:
To ensure that only 'Femenino' and 'Masculino' categories are included in the analysis as per the user's request, I will filter the DataFrame to exclude all entries where 'genero_director' is 'Desconocido'. This step will be performed before grouping the data by year and gender.

In [36]:
import pandas as pd
import os

file_path = "/content/cine-chile-limpia-oficial.csv"

# Check if the file exists
if not os.path.exists(file_path):
    print(f"Error: El archivo '{file_path}' no se encontró. Por favor, asegúrate de subir el archivo CSV a esta ubicación.")
else:
    # Load df if it's not already loaded or is empty, assuming file exists
    if 'df' not in locals() or df.empty:
        df = pd.read_csv(file_path, sep=";")

    # Define gender inference components if not already defined
    if 'nombres_masculinos' not in locals() or 'nombres_femeninos' not in locals() or 'inferir_genero' not in locals():
        nombres_masculinos = [
            'José', 'Juan', 'Luis', 'Carlos', 'Francisco', 'Manuel', 'Antonio', 'Miguel', 'Pedro', 'Pablo',
            'Fernando', 'Alejandro', 'Andrés', 'David', 'Jorge', 'Daniel', 'Javier', 'Sergio', 'Ricardo', 'Alberto',
            'Eduardo', 'Enrique', 'Rafael', 'Diego', 'Gonzalo', 'Héctor', 'Mario', 'Martín', 'Oscar', 'Ramón',
            'Roberto', 'Rubén', 'Santiago', 'Sebastián', 'Vicente', 'Esteban', 'Álvaro', 'Gabriel', 'Guillermo', 'Ignacio',
            'Joaquín', 'Leonardo', 'Marcos', 'Nicolás', 'Raúl', 'Rodrigo', 'Adrián', 'Agustín', 'Alfredo', 'Benjamín',
            'Cristián', 'Felipe', 'Germán', 'Gustavo', 'Patricio', 'Víctor'
        ]
        nombres_femeninos = [
            'María', 'Ana', 'Laura', 'Sofía', 'Valentina', 'Camila', 'Isabel', 'Paula', 'Daniela', 'Carolina',
            'Andrea', 'Elena', 'Gabriela', 'Jimena', 'Lucía', 'Mariana', 'Natalia', 'Pilar', 'Sara', 'Silvia',
            'Verónica', 'Alejandra', 'Beatriz', 'Carmen', 'Clara', 'Diana', 'Dolores', 'Emilia', 'Eva', 'Florencia',
            'Gloria', 'Irene', 'Julia', 'Lorena', 'Magdalena', 'Marta', 'Mónica', 'Patricia', 'Raquel', 'Rocío',
            'Rosa', 'Sandra', 'Teresa', 'Victoria', 'Ximena', 'Cecilia', 'Consuelo', 'Francisca', 'Jessica', 'Katherine',
            'Claudia', 'Elizabeth', 'Macarena', 'Melissa', 'Romina', 'Soledad'
        ]

        def inferir_genero(nombre_completo):
            if pd.isna(nombre_completo):
                return 'Desconocido'
            nombre_completo_lower = str(nombre_completo).lower()
            primer_nombre = nombre_completo_lower.split(' ')[0]
            for nombre_fem in nombres_femeninos:
                if primer_nombre == nombre_fem.lower():
                    return 'Femenino'
            for nombre_masc in nombres_masculinos:
                if primer_nombre == nombre_masc.lower():
                    return 'Masculino'
            return 'Desconocido'
        df['genero_director'] = df['director-01'].apply(inferir_genero)

    # Filter out 'Desconocido' entries if df is available
    if 'df' in locals() and not df.empty:
        df_filtrado = df[df['genero_director'] != 'Desconocido'].copy() # Use .copy() to avoid SettingWithCopyWarning
        print("DataFrame filtrado para excluir directores 'Desconocido'.")
        print(df_filtrado['genero_director'].value_counts())

        # Re-calculate directores_por_ano_genero using the filtered DataFrame
        directores_por_ano_genero = df_filtrado.groupby(['año', 'genero_director'])['director-01'].nunique().reset_index()

        print("Conteo de directores únicos por año y género (solo Femenino/Masculino):")
        print(directores_por_ano_genero.head())
    else:
        print("No se pudo cargar el DataFrame 'df', no se realizó el filtrado ni el conteo.")

DataFrame filtrado para excluir directores 'Desconocido'.
genero_director
Masculino    1106
Femenino      145
Name: count, dtype: int64
Conteo de directores únicos por año y género (solo Femenino/Masculino):
    año genero_director  director-01
0  1915       Masculino            1
1  1917        Femenino            1
2  1917       Masculino            1
3  1919       Masculino            1
4  1920       Masculino            2


In [39]:
if 'directores_pivot' in locals():
    directores_pivot['ratio_F_M'] = directores_pivot['Femenino'] / directores_pivot['Masculino']
    directores_pivot['ratio_F_M'] = directores_pivot['ratio_F_M'].replace([float('inf'), -float('inf')], 0).fillna(0)

    print("DataFrame con la tasa de mujeres por hombres (solo Femenino/Masculino):")
    print(directores_pivot.head())
else:
    print("No se ha definido 'directores_pivot', no se puede calcular la tasa F/M.")

DataFrame con la tasa de mujeres por hombres (solo Femenino/Masculino):
genero_director   año  Desconocido  Femenino  Masculino  ratio_F_M
0                1915          0.0       0.0        1.0        0.0
1                1916          1.0       0.0        0.0        0.0
2                1917          1.0       1.0        1.0        1.0
3                1918          1.0       0.0        0.0        0.0
4                1919          0.0       0.0        1.0        0.0


**Reasoning**:
With the 'Desconocido' category removed, I will now re-calculate the `directores_por_ano_genero` DataFrame using the filtered data. This will ensure that all subsequent calculations and visualizations only consider 'Femenino' and 'Masculino' directors.

## Final Task

### Subtask:
Se presentará la visualización final y se discutirán los patrones observados en la evolución de la participación de directores y directoras, así como la tasa de mujeres por hombre a lo largo de los años.


## Summary:

### Data Analysis Key Findings

*   **Gender Classification:** From the initial classification, 1,106 directors were identified as 'Masculino', 145 as 'Femenino', and 868 remained 'Desconocido' due to limitations in the name list.
*   **Annual Director Participation:** The analysis successfully generated annual counts of unique directors by inferred gender, showing their evolution over time in a stacked area chart.
*   **Female-to-Male Ratio:** A `ratio_F_M` was calculated for each year, indicating the number of female directors per male director. Years with zero male directors were handled by setting the ratio to 0.
*   **Visual Trend Representation:** A combined Altair chart was created, effectively visualizing the stacked area of gender distribution and overlaying the female-to-male ratio line, demonstrating the evolution of gender representation in Chilean cinema over the years.

### Insights or Next Steps

*   **Improve Gender Inference:** The high number of 'Desconocido' directors (868) suggests that the current name lists are insufficient. Expanding the `nombres_masculinos` and `nombres_femeninos` lists, or incorporating more sophisticated name-to-gender inference tools (e.g., using external APIs or larger gender-name databases), could significantly improve the accuracy of gender classification.
*   **Analyze Ratio Fluctuations:** Further investigation into specific years where the female-to-male ratio shows significant spikes or drops could reveal underlying social, economic, or industry-specific events that influenced gender representation in film direction during those periods.
